pkg/controllers/util/federatedclient/podinformer.go

```diff
@@ -0,0 +1,123 @@
+package federatedclient
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/informers"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+func addPodInformer(ctx context.Context,
+	informer informers.SharedInformerFactory,
+	client kubeclient.Interface,
+	podListerSemaphore *semaphore.Weighted,
+	enableSimplifying bool) {
+	informer.InformerFor(&corev1.Pod{}, func(k kubeclient.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
+		return cache.NewSharedIndexInformer(
+			podListerWatcher(ctx, client, podListerSemaphore, enableSimplifying),
+			&corev1.Pod{},
+			resyncPeriod,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		)
+	})
+}
+
+func podListerWatcher(ctx context.Context, client kubeclient.Interface, semaphore *semaphore.Weighted, enableSimplifying bool) cache.ListerWatcher {
+	return &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			if semaphore != nil {
+				if err := semaphore.Acquire(ctx, 1); err != nil {
+					return nil, err
+				}
+				defer semaphore.Release(1)
+			}
+			pods, err := client.CoreV1().Pods(corev1.NamespaceAll).List(ctx, options)
```

Comment:
The client-go informers use `context.TODO()`. Intuitively I think it does make sense to shutdown the listwatch when the federatedclient is shutting down, but I can't be sure. @JackZxj @limhawjia @SOF3 @mrlihanbo WDYT?

Reply:
I think we can reorganize all context usages in a separate PR; the scope is too big and doesn't need to be implemented here.

Reply:
Agree that context usage should be fixed in a separate PR. I'm just wondering whether using `ctx` instead of `context.TODO` here is correct. Using `ctx` would mean the `ListWatch` would shutdown when `ctx` is cancelled.